tls: add option to set signature algorithms

tsaarni opened the pull request:
Add a new optional field `signature_algorithms` to `TlsParameters` message to configure signature algorithms. The value is used during TLS handshake in TLS Signature Algorithms extension. It indicates which signature algorithms may be used for digital signatures. The value can be used for both `DownstreamTlsContext` and `UpstreamTlsContext`.

When user does not set the option, the default value is defined by BoringSSL (like it has been before this PR). 

Fixes #26166

**Additional Description:**

There can be use cases that calls for exact control over the used signature algorithms. For example, legacy algorithm such as `rsa_pkcs1_sha1` may be [discouraged](https://www.ietf.org/archive/id/draft-mattsson-tls-psk-ke-dont-dont-dont-05.html#name-tls-signaturescheme) and user might want to disable it. This PR gives user the option to override the default list of algorithms.

**Risk Level:** Medium

**Testing:**

New test case is added to verify the value of TLS Signature Algorithms extension during TLS handshake. Test case does not utilize the usual `testUtil/testUtilV2` flow since it needs to plug into `SSL_set_cert_cb()` callback to get access to full list of signature algorithms offered by peer during TLS handshake.

**Docs Changes:**

New field documented in `.proto`.

**Release Notes:**

tls: added new field `TlsParameters.signature_algorithms` to set signature algorithms.

ggreenway reviewed (changes requested):
This feature makes sense, but I think it is preferable to leave the default as referencing the boringssl default, when the new setting is unspecified. The rationale is that if boringssl changes its defaults for security reasons, I'd rather not need any additional changes in envoy to pick up that change, other than updating to a newer version of boringssl. 

/wait

tsaarni commented:
> This feature makes sense, but I think it is preferable to leave the default as referencing the boringssl default, when the new setting is unspecified. The rationale is that if boringssl changes its defaults for security reasons, I'd rather not need any additional changes in envoy to pick up that change, other than updating to a newer version of boringssl.

Fixed.

With this change I also changed Risk Level from "_Medium / High_" to "_Medium_" since now by default there should be no impact from this change.

ggreenway commented on source/extensions/transport_sockets/tls/context_impl.cc:
```diff
@@ -157,6 +157,13 @@ ContextImpl::ContextImpl(Stats::Scope& scope, const Envoy::Ssl::ContextConfig& c
     }
+
+    // Set signature algorithms if given, otherwise fall back to BoringSSL defaults.
+    if (!capabilities_.provides_sigalgs && !config.signatureAlgorithms().empty() &&
```
nit: for readability, break up the condition for whether to call `SSL_CTX_set1_sigalgs_list` from the error handling for a failure in that call.

ggreenway reviewed (changes requested):
This LGTM overall

/wait

ggreenway reviewed (dismissed):
LGTM, but there are some CI failures

/wait-any

tsaarni commented:
@ggreenway I've added a test case to cover the error branch as well. There are only few changes in non-test files and I don't know what else could be improved to get coverage test passed. 

Various other tests have failed during the test runs, but I'm not sure what the connection is to the change. I've now merged in case some flaky tests have been fixed in main since I started. Now I see `//test/extensions/filters/http/tap:tap_filter_integration_test` failing ([link](https://dev.azure.com/cncf/envoy/_build/results?buildId=132205&view=logs&j=bbe4b42d-86e6-5e9c-8a0b-fea01d818a24&t=00ade0d6-ffae-5af0-b5e0-e33b1813c8bc)), possibly related to [existing problems](https://github.com/envoyproxy/envoy/issues/25663). Also, some packaging tests fail, possibly to download errors ([link](https://dev.azure.com/cncf/envoy/_build/results?buildId=132205&view=logs&j=17c773a2-2e8d-5b7d-1064-faa2b1aabe48&t=3c4ae1aa-54bd-5ebd-2552-b3b4aa8e453b)).

ggreenway commented:
/retest

ggreenway reviewed (changes requested):
Please add a release note, then this should be all set.

/wait
